### PR TITLE
[FLINK-24614][Connectors/PARQUET] Add Complex types support for parquet reader

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change

Add array,map,row types support for parquet vectorized reader

## Brief change log

  - *Add array,map,row types support for parquet vectorized reader*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests in `ParquetColumnarRowSplitReaderTest` and `ParquetColumnarRowInputFormatTest`*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no )

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@flinkbot run azure

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReaderFactory.java [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+ * Parquet file has self-describing schema which may differ from the user required schema (e.g.
+ * schema evolution). This factory is used to retrieve user required typed data via corresponding
+ * reader which reads the underlying data.
+ */
+public final class ParquetDataColumnReaderFactory {
+
+    private ParquetDataColumnReaderFactory() {}
+
+    /** default reader for {@link ParquetDataColumnReader}. */
+    public static class DefaultParquetDataColumnReader implements ParquetDataColumnReader {
```
Comment (luoyuxia) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReaderFactory.java:46]: This class mix Dictionary reader and ValuesReader, which cause many warning like that`Method invocation 'readBoolean' may produce 'NullPointerException`  in idea 
Would it be better to split it into two class, one for ValuesReader and one for Dictionary reader?

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java [hunk 2]
```diff
@@ -85,45 +91,43 @@ public class ParquetColumnarRowInputFormatTest {
...
+                    new DecimalType(15, 0),
+                    new DecimalType(20, 0),
+                    new ArrayType(new VarCharType(VarCharType.MAX_LENGTH)),
+                    new ArrayType(new BooleanType()),
+                    new ArrayType(new TinyIntType()),
+                    new ArrayType(new SmallIntType()),
+                    new ArrayType(new IntType()),
+                    new ArrayType(new BigIntType()),
+                    new ArrayType(new FloatType()),
+                    new ArrayType(new DoubleType()),
```
Comment (luoyuxia) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java:118]: Add a test like `new ArrayType(new ArrayType(new DoubleType()))`?

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java [hunk 3]
```diff
@@ -85,45 +91,43 @@ public class ParquetColumnarRowInputFormatTest {
...
+                    new VarCharType(VarCharType.MAX_LENGTH),
+                    new BooleanType(),
+                    new TinyIntType(),
+                    new SmallIntType(),
+                    new IntType(),
+                    new BigIntType(),
+                    new FloatType(),
+                    new DoubleType(),
+                    new TimestampType(9),
+                    new DecimalType(5, 0),
```
Comment (luoyuxia) [inline flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java:105]: also test DecimalType with scale like DecimalType(5, 2)?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/columnar/vector/heap/AbstractHeapVector.java [hunk 4]
```diff
@@ -48,8 +48,11 @@ public abstract class AbstractHeapVector extends AbstractWritableVector {
     /** Reusable column for ids of dictionary. */
     protected HeapIntVector dictionaryIds;
 
+    private int len;
```
Comment (luoyuxia) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/columnar/vector/heap/AbstractHeapVector.java:51]: nit:
```suggestion
    private final int len;
```

**Review by luoyuxia — COMMENTED** [review]
@lsyldliu The pr looks generally good to me. I left some comments.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReader.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    int readTinyInt();
+
+    /** @return the next Float from the page */
+    float readFloat();
+
+    /** @return the next Boolean from the page */
+    boolean readBoolean();
+
+    /** @return the next String from the page */
+    byte[] readString();
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReader.java:63]: Here doesn't exist `StringVector`, so we can't return string directly.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java [hunk 6]
```diff
@@ -396,6 +490,48 @@ public static WritableColumnVector createWritableColumnVector(
                             typeName);
                     return new HeapBytesVector(batchSize);
                 }
+            case ARRAY:
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:493]: Due to `PrimitiveType` only support non-nested column, so we can't check nested column. BTW, the element type of nested type will be checked recursively, so I think check primitive type is enough.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java [hunk 7]
```diff
@@ -257,56 +271,136 @@ public static ColumnVector createVectorFromConstant(
...
             LogicalType fieldType,
-            ColumnDescriptor descriptor,
-            PageReader pageReader)
+            Type type,
+            List<ColumnDescriptor> columnDescriptors,
+            PageReadStore pages,
+            int depth)
             throws IOException {
+        List<ColumnDescriptor> descriptors =
+                getAllColumnDescriptorByType(depth, type, columnDescriptors);
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/ParquetSplitReaderUtil.java:297]: Regarding to arbitrary type, the `ColumnDescriptor` should exists, so I have check the list in `getAllColumnDescriptorByType `.

File: flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java (hunk 2 above)
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/test/java/org/apache/flink/formats/parquet/ParquetColumnarRowInputFormatTest.java:118]: Parquet writer doesn't support writer `new ArrayType(new ArrayType(new DoubleType()))` type now, so this type of test is not good to add, please refer to `ArrayWriter`.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReaderFactory.java (hunk 1 above)
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ParquetDataColumnReaderFactory.java:46]: I think your advice is very good, although split it into two class maybe seem to more clear, but after read hive related code, I think we can refer to hive. I have remove the @Nullable annotation.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+                return dataColumn.readDouble();
+            case DECIMAL:
+                switch (descriptor.getPrimitiveType().getPrimitiveTypeName()) {
+                    case INT32:
+                        return dataColumn.readInteger();
+                    case INT64:
+                        return dataColumn.readLong();
+                    case BINARY:
+                    case FIXED_LEN_BYTE_ARRAY:
+                        return dataColumn.readString();
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/ArrayColumnReader.java:170]: I've remove the unused method currently.

**Comment by lsyldliu** [comment]
@luoyuxia Thanks for review, I've addressed comments.

**Comment by lsyldliu** [comment]
@flinkbot run azure

**Review by wuchong — COMMENTED** [review]
I just have one question that does this PR only supports ARRAY, MAP, ROW with primitive types? Does parquet support nested types in array, map, row?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/BaseVectorizedColumnReader.java [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+import static org.apache.parquet.column.ValuesType.DEFINITION_LEVEL;
+import static org.apache.parquet.column.ValuesType.REPETITION_LEVEL;
+import static org.apache.parquet.column.ValuesType.VALUES;
+
+/**
+ * It's column level Parquet reader which is used to read a batch of records for a column, part of
+ * the code is referred from Apache Hive and Apache Parquet.
+ */
+public abstract class BaseVectorizedColumnReader implements ColumnReader<WritableColumnVector> {
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/BaseVectorizedColumnReader.java:51]: Yes, I think it is possible. Due to the time constraint, I can't complete the combine work now, so I will refactor it in next version.

**Comment by lsyldliu** [comment]
> I just have one question that does this PR only supports ARRAY, MAP, ROW with primitive types? Does parquet support nested types in array, map, row?

1. For the first question, yes for array and map, row can nested with map and array.
2. For the second question, it seems that parquet support nested types in nested type, refer to https://parquet.apache.org/docs/file-format/nestedencoding/

After investigate [hive](https://github.com/apache/hive/blob/27dd6fdc89b10c8813c02310291b09565f0d4959/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedListColumnReader.java#L38) and [spark](https://github.com/apache/spark/blob/35d00df9bba7238ad4f409999617fae4d04ddbfd/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java#L260) implementation, I found hive doesn't support array with nested types currently, but spark support. So I think we should assign with spark in next version.

**Comment by lsyldliu** [comment]
Thanks for all review, I've addressed comments.

**Comment by wuchong** [comment]
> After investigate [hive](https://github.com/apache/hive/blob/27dd6fdc89b10c8813c02310291b09565f0d4959/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedListColumnReader.java#L38) and [spark](https://github.com/apache/spark/blob/35d00df9bba7238ad4f409999617fae4d04ddbfd/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java#L260) implementation, I found hive doesn't support array with nested types currently, but spark support. So I think we should assign with spark in next version.

Thank you. Could you create an issue for that?

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by lsyldliu** [comment]
> > After investigate [hive](https://github.com/apache/hive/blob/27dd6fdc89b10c8813c02310291b09565f0d4959/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/vector/VectorizedListColumnReader.java#L38) and [spark](https://github.com/apache/spark/blob/35d00df9bba7238ad4f409999617fae4d04ddbfd/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/ParquetColumnVector.java#L260) implementation, I found hive doesn't support array with nested types currently, but spark support. So I think we should assign with spark in next version.
> 
> Thank you. Could you create an issue for that?

Okay, I will create it.

**Comment by lsyldliu** [comment]
> 

https://issues.apache.org/jira/browse/FLINK-28867

**Pull request merged**